### DISCUSSION TITLE
Effect of Pruning on inference cost

### DISCUSSION THREAD

**Discussion by shahram-WINS** [Xilinx/brevitas-radioml-challenge-21#16] [Export & Inference Cost]
Hello, when I want to calculate the cost on my laptop, I follow the first couple of steps in the notebook to update the model, and then I jump to the last four steps to calculate the inference cost. We could also calculate the cost by using the trained_model file and following the steps after the "Load a Trained Model" section in the notebook. Generally, the obtained inference cost by any of these two techniques should be the same. However, when I use the pruning technique for reducing the cost in the model, the values of the inference cost are varied (the inference cost calculated by the trained_model would be higher). 

So I wonder what could affect the inference cost calculation when pruning is used in the model, and how you validate the inference cost when a pruning technique is used?

**Comment by maltanar** [comment]
Hi @shahram-WINS , I'm not sure if I understood your question, especially this part:

> However, when I use the pruning technique for reducing the cost in the model, the values of the inference cost are varied (the inference 
 cost calculated by the trained_model would be higher).

If you prune weights from the trained model, you need to make sure you export that to a different ONNX file and call `inference_cost` on that file instead.

Perhaps explaining the mechanichs of how we take sparsity into account will help. When you call the `inference_cost` function on the ONNX model, it will start by finding the convolutional and FC layers in the model. For each of those layers, the cost calculation will look at [weight density](https://github.com/Xilinx/finn-base/blob/feature/itu_competition_21/src/finn/analysis/inference_cost.py#L111) when `discount_sparsity=True` for that layer's weights. In turn, you can see how the weight density is computed [here](https://github.com/Xilinx/finn-base/blob/311ea084b01322f2959bdc35c12f70b638c9367c/src/finn/analysis/inference_cost.py#L63): it looks at how many weight values are equal to zero compared to how many weights there are, and returns a number between 0.0 and 1.0 depending on that ratio.
